### ISSUE TITLE
Exception handling bug exposed with System.Xml.RW.WrappedReader.Tests.dll

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#633]
@janvorli is investigating; it's difficult to create an isolated repro, but some tests are failing in System.Xml.RW.WrappedReader.Tests.dll (and a few other similar XML reader writer test projects) due to what appears to be nested catch blocks not catching exceptions they should be catching.

**Comment by MattWhilden** [comment]
I presume this is for xplat... or is it omnipresent?

**Comment by stephentoub** [comment]
Sorry, yes, on Unix.

**Comment by kangaroo** [comment]
@stephentoub I'm going to hijack this issue slightly.  You've filed a few on tests failing for xplat.  I apologize if this is covered somewhere else, but are their build/run instructions for xplat xunit tests yet?  I've got a day of travel on Monday, and would like to check the current state for OS X on my flight

**Comment by MattWhilden** [comment]
@jkotas May also be able to hook you up with xplat test run instructions.

**Comment by stephentoub** [comment]
@kangaroo, thanks!  We'd welcome any assistance here.  I've been very focused on first getting things going on Linux, and haven't spent much time working through the tests on Mac yet, so any assistance there would be very welcome.  I'm confident there will be issues to fix ;)

Unfortunately the testing process on Linux and Mac are still rudimentary, in large part due to still only building the managed libraries / tests on Windows. It'll get better soon, but for the moment it requires manual efforts.

Actually running the tests is done using the same command that msbuild / build on Windows kicks off.  Given a test assembly like System.Threading.Tasks.Dataflow.Tests.dll, you run the command:
```
./corerun xunit.console.netcore.exe System.Threading.Tasks.Dataflow.Tests.dll additionalArgsToBeExplainedShortly
```
The xunit.console.netcore.exe here is the same one used on Windows, and you can fish that out of either the build tools in your packages directory on Windows, or you can navigate through to find it in any one of the test directories, e.g.
```\corefx\bin\tests\Windows_NT.AnyCPU.Debug\System.Threading.Tasks.Dataflow.Tests\aspnetcore50\xunit.console.netcore.exe```
That's also where you'll find the Test assembly:
```"C:\Users\stoub\Source\Repos\corefx\bin\tests\Windows_NT.AnyCPU.Debug\System.Threading.Tasks.Dataflow.Tests\aspnetcore50\System.Threading.Tasks.Dataflow.Tests.dll"````

I typically do a full build on Windows and then copy out of ```bin\tests``` all ```*.Tests.dll``` that live in any subdirectory.  I copy those to my Linux machine into the same folder that has the runtime and the rest of the assemblies.  A few of the libraries have some other files that need to be copied as well, e.g. there's another test app that gets built as part of the process tests.  You can copy the xunit.* files from any of the aspnetcore50 directories under test... they're all the same, so you really only need to do that once.

Once you have all that, you can run tests.  I noted above that there are some additional arguments needed.  We have a traits system with the unit tests, where tests can be annotated with traits and then you can tell xunit to only run those with or without certain traits.  There are some double-negatives here that make it a bit confusing, but on OSX you'd want to add:
```-notrait category=nonosxtests```
meaning "run any tests that don't have the trait 'don't run this on OSX'" :smile:  You also want to tell it to only run the "inner loop" tests; those are the same subset run during PR CI builds (at least for Windows; coming soon for Linux and Mac).  That's achieved with:
```-notrait category=outerloop```
So, to run the inner loop dataflow tests noted as being ok to run on the Mac, do:
```
./corerun xunit.console.netcore.exe System.Threading.Tasks.Dataflow.Tests.dll -notrait category=nonosxtests -notrait category=outerloop
```

If I've updated the product library, I of course copy that as well, but if it's a library that has an OS-specific implementation, I need to copy the right one.  For example, whereas System.Threading.Tasks.Dataflow.dll will be the same regardless of OS, System.Console.dll has a custom implementation for each OS.  So, you'd do "build /p:OS=OSX" on Windows, and that'll generate the necessary libraries under ```bin\OSX.AnyCpu.Debug```, which you can then copy over.

I hope that helps, and sorry this was a bit rambly.  We'll have a better process soon, including instructions.

**Comment by kangaroo** [comment]
Thanks @stephentoub -- thats not too bad.  I'll sync up and see where things are before my flight.

**Comment by janvorli** [comment]
Fixed by #660
